src/embed/http.rs

```diff
@@ -0,0 +1,92 @@
+use js_sys;
+use std::convert::TryFrom;
+use std::error::Error;
+use std::fmt;
+use std::str;
+use wasm_bindgen::JsCast;
+use wasm_bindgen::JsValue;
+use wasm_bindgen_futures::JsFuture;
+use web_sys::{RequestInit, RequestMode};
+
+// browser_fetch makes and HTTP request over the network via the browser's Fetch API.
+// It is intended to be used in wasm; it won't work in regular rust. The response contains
+// the status and body, it doesn't populate response headers (though it easily could).
+// Since we can't run a web server in wasm land and this fn doesn't work in rust land
+// this fn is tested manually in tests/wasm.rs.
+//
+// TODO timeout/abort
+// TODO log request/response
+pub async fn browser_fetch(
+    http_req: &http::Request<String>,
+) -> Result<http::Response<String>, FetchError> {
+    let mut opts = RequestInit::new();
+    opts.method(http_req.method().as_str());
+    opts.mode(RequestMode::Cors);
+    let js_body = JsValue::from_str(http_req.body());
+    opts.body(Some(&js_body));
+    let web_sys_req = web_sys::Request::new_with_str_and_init(&http_req.uri().to_string(), &opts)?;
+    let h = web_sys_req.headers();
+    for (k, v) in http_req.headers().iter() {
+        h.set(k.as_ref(), v.to_str().unwrap())?;
+    }
+
+    let window = web_sys::window().ok_or(FetchError::new("could not get window"))?;
+    let http_req_promise = window.fetch_with_request(&web_sys_req);
+    let http_req_future = JsFuture::from(http_req_promise);
+    let js_web_sys_resp = http_req_future.await?;
+    if !js_web_sys_resp.is_instance_of::<web_sys::Response>() {
+        return Err(FetchError::new("result from future not a Response"));
+    }
+    let web_sys_resp: web_sys::Response = js_web_sys_resp.dyn_into().unwrap();
+    let body_js_value = JsFuture::from(web_sys_resp.text()?).await?;
+    let resp_body = body_js_value.as_string().ok_or(FetchError::new(
+        "could not get http response body as string",
+    ))?;
+
+    let builder = http::response::Builder::new();
+    let http_resp = builder.status(web_sys_resp.status()).body(resp_body)?;
+    Ok(http_resp)
+}
+
+// FetchErrors are returned by both the rust and browser versions of fetch. Since
+// lower level errors in each case will be coming from two different places I implemented
+// FetchError as lossy of the error types underneath: it holds an error string.
+#[derive(Debug)]
+pub struct FetchError {
```

Comment:
i know i can do this with `Enumtype(String)` without implementing Error and that elsewhere we are enumerating error cases as opposed to using strings to describe them. i did this first before i understood that. my sense is that if you'll want us to standardize around enumerated types rather than what we have here or FetchError(String).

Reply:
I really like the enumerated types and think we could create machinery to make them easier to use. What do you think?

Reply:
1. can't decide whether i'm annoyed at having all the fidelity i don't care about versus reassured that each case is doing the right thing. i liked the go style of only formalizing that which the caller would wants to treat like a first class thing. but enumerating all cases maybe does seem more rusty. 
2. i like human-readable error strings that carry the bits of information i'm interested, when such bits exist. ParseError is a lot less useful than ParseError("unexpected token 'noll' attempting to parse as json '{"foo": noll}'"). so as long as we can have that sounds good.

i dont have a great deal of experience and the enumerated types seem pretty lightweight so i'm game for giving it a go.